reposerver/repository/repository_test.go

```diff
@@ -202,6 +203,305 @@ func TestGenerateHelmChartWithDependencies(t *testing.T) {
+				}
+
+				isCachedError := err != nil && strings.HasPrefix(err.Error(), cachedManifestGenerationPrefix)
+
+				if adjustedInvocation < service.initConstants.PauseGenerationAfterFailedGenerationAttempts {
+					// GenerateManifest should not return cached errors for the first X responses, where X is the FailGenAttempts constants
+					assert.True(t, !isCachedError)
+					assert.True(t, manifestTestUtil.recentCachedRes != nil)
+					assert.True(t, manifestTestUtil.recentCachedRes.FirstFailureTimestamp != 0)
+
+					// Internal cache consec failures value should increase with invocations, cached response should stay the same,
+					assert.True(t, int(manifestTestUtil.recentCachedRes.NumberOfConsecutiveFailures) == adjustedInvocation+1)
+					assert.True(t, int(manifestTestUtil.recentCachedRes.NumberOfCachedResponsesReturned) == 0)
+
+				} else {
+					// GenerateManifest SHOULD return cached errors for the next X responses, where X is the
+					// PauseGenerationOnFailureForRequests constant
+					assert.True(t, isCachedError)
+					assert.True(t, manifestTestUtil.recentCachedRes != nil)
+					assert.True(t, manifestTestUtil.recentCachedRes.FirstFailureTimestamp != 0)
+
+					// Internal cache values should update correctly based on number of return cache entries, concecutive failures should stay the same
+					assert.True(t, int(manifestTestUtil.recentCachedRes.NumberOfConsecutiveFailures) == service.initConstants.PauseGenerationAfterFailedGenerationAttempts)
+					assert.True(t, int(manifestTestUtil.recentCachedRes.NumberOfCachedResponsesReturned) == (adjustedInvocation-service.initConstants.PauseGenerationAfterFailedGenerationAttempts+1))
+				}
+
+			}
+
+		})
+	}
+
+}
+
+func TestManifestGenErrorCacheFileContentsChange(t *testing.T) {
+
+	tmpDir, err := ioutil.TempDir("", "repository-test-")
+	assert.NoError(t, err)
+	defer os.RemoveAll(tmpDir)
+
+	service := newService(tmpDir)
+
+	service.initConstants = RepoServerInitConstants{
+		ParallelismLimit: 1,
+		PauseGenerationAfterFailedGenerationAttempts: 2,
+		PauseGenerationOnFailureForMinutes:           0,
+		PauseGenerationOnFailureForRequests:          4,
+	}
+
+	for step := 0; step < 3; step++ {
+
+		// step 1) Attempt to generate manifests against invalid helm chart (should return uncached error)
+		// step 2) Attempt to generate manifest against valid helm chart (should succeed and return valid response)
+		// step 3) Attempt to generate manifest against invalid helm chart (should return cached value from step 2)
+
+		errorExpected := step%2 == 0
+
+		// Ensure that the target directory will succeed or fail, so we can verify the cache correctly handles it
+		err = os.RemoveAll(tmpDir)
+		assert.NoError(t, err)
+		err = os.MkdirAll(tmpDir, 0777)
+		assert.NoError(t, err)
+		if errorExpected {
+			// Copy invalid helm chart into temporary directory, ensuring manifest generation will fail
+			err = fileutil.CopyDir("./testdata/invalid-helm", tmpDir)
```

Comment:
Since we're in a test, I'm unsure whether an exec of `cp -r ...` would suffice for this. Also, I believe there are couple of pure-go implementations for it already. I'm not objecting to the current approach however, just loudly thinking :)

Reply:
Hey @jannfis, would you be able to provide more info on this review comment? I wasn't sure if you were objecting to the specific implementation of `fileutil.CopyDir(...)`, or you were objecting to the idea of creating a temporary directory and copying files into it as part of a unit test.

In either case, here's a bit more information on usage:
- My PR does include a pure Go `CopyDir(...)` function for copying directory contents (rather than calling `cp -r` directly, which would not, for example, work on Windows).
- The use of the Go-generated OS temporary folder should ensure that the files are cleaned up (if, in rare cases, the defer delete function wasn't called). The generated temporary folder names contain sufficient entropy that if multiple similar unit tests are running concurrently then they should not interfere with each other.
- The existing `repository_test.go` unit tests' call a live version of `helm` from the filesystem via the CLI, rather than a mocking the `helm` call. This PR's test code works the same way, and thus needs an actual copy of an (invalid) helm chart on the file system for the actual Helm CLI to use, hence the need for a folder copy operation.

Reply:
Hey, sorry if I was a little unclear in my phrasing. 

Actually I was thinking whether using `cp -r` would be sufficient in this use-case (copying files within a unit test). While I agree that this would not work on Windows and other non-Unix like systems, I also think that none of the build process currently does either :)

Anyway, if we keep the code in `util/files/files.go`, I guess it will encourage people to use it for future code. In this case, we should also provide dedicated unit tests for it IMHO - or we should make it clear that it's only used in tests, and therefore move it somewhere as a fixture-like function (i.e. somewhere at `test/fixture/...`).

Did that make a little more sense? :thinking:

Reply:
Excellent, thanks, I definitely thought of this bit of code as 'for tests only' rather than something that would be used by the product code itself, so I've moved it into `test/fixture/path/...`, which seems like a good spot for it.